examples/platforms/utils/mac_frame.cpp

```diff
@@ -0,0 +1,93 @@
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "mac_frame.h"
+
+#include "mac/mac_frame.hpp"
+
+using namespace ot;
+
+bool otMacFrameIsMatched(const otRadioFrame *aFrame,
+                         otPanId             aPanId,
+                         otShortAddress      aShortAddress,
+                         const otExtAddress *aExtAddress)
+{
+    const Mac::Frame &frame = *static_cast<const Mac::Frame *>(aFrame);
+    bool              rval  = false;
+    Mac::Address      dst;
+    Mac::PanId        panid;
+
+    SuccessOrExit(frame.GetDstAddr(dst));
+
+    switch (dst.GetType())
+    {
+    case Mac::Address::kTypeShort:
+        SuccessOrExit(frame.GetDstPanId(panid));
+        rval = (panid == Mac::kPanIdBroadcast || panid == aPanId) &&
+               (dst.GetShort() == Mac::kShortAddrBroadcast || dst.GetShort() == aShortAddress);
+        break;
+
+    case Mac::Address::kTypeExtended:
+        SuccessOrExit(frame.GetDstPanId(panid));
+        rval = (panid == Mac::kPanIdBroadcast || panid == aPanId) &&
+               dst.GetExtended() == *static_cast<const Mac::ExtAddress *>(aExtAddress);
+        break;
+
+    case Mac::Address::kTypeNone:
+        rval = true;
+        break;
+    }
+
+exit:
+    return rval;
+}
+
+bool otMacFrameIsAck(const otRadioFrame *aFrame)
+{
+    return static_cast<const Mac::Frame *>(aFrame)->GetType() == Mac::Frame::kFcfFrameAck;
+}
+
+bool otMacFrameIsDataRequest(const otRadioFrame *aFrame)
+{
+    return static_cast<const Mac::Frame *>(aFrame)->IsDataRequestCommand();
+}
+
+bool otMacFrameIsAckRequested(const otRadioFrame *aFrame)
+{
+    return static_cast<const Mac::Frame *>(aFrame)->GetAckRequest();
+}
+
+otError otMacFrameGetSrcAddr(const otRadioFrame *aFrame, otMacAddress *aMacAddress)
+{
+    return static_cast<const Mac::Frame *>(aFrame)->GetSrcAddr(*reinterpret_cast<Mac::Address *>(aMacAddress));
```

Comment:
Do we know for sure that the `reinterpret_cast<Mac::Address *>` would work on all toolchains/compliers?
To be safe, how about we add code to convert, something like this:
```
otError otMacFrameGetSrcAddr(const otRadioFrame *aFrame, otMacAddress *aMacAddress)
{
	otError error;
	Mac::Address address;

	error = static_cast<const Mac::Frame *>(aFrame)->GetSrcAddr(address);
	SuccessOrExit(error);

	switch (address.GetType())
	{
	case Mac::Address::kTypeNone:
		aMacAddress->mType = OT_MAC_ADDRESS_TYPE_NONE;
		break;

	case Mac::Address::kTypeShort:
	    aMacAddress->mType = OT_MAC_ADDRESS_TYPE_SHORT;
		aMacAddress->mAddress.mShortAddress = address.GetShort();
		break;

	case Mac::Address::kTypeExtended:
	    aMacAddress->mType = OT_MAC_ADDRESS_TYPE_EXTENDED
		aMacAddress->mAddress.mExtendedAddress = address.GetExtended();
		break;
	}

exit:
	return error;
}
```